src/operations/SymbolicOperation.cpp

```diff
@@ -0,0 +1,286 @@
+        SymbolicOperation(nq, targets, g, lambda, phi, theta, startingQubit) {
+        this->controls = controls;
+    }
+
+    // MCF (cSWAP), Peres, paramterized two target Constructor
+    SymbolicOperation::SymbolicOperation(dd::QubitCount nq, const dd::Controls& controls, dd::Qubit target0, dd::Qubit target1, OpType g, const SymbolOrNumber& lambda, const SymbolOrNumber& phi, const SymbolOrNumber& theta, dd::Qubit startingQubit):
+        SymbolicOperation(nq, controls, {target0, target1}, g, lambda, phi, theta, startingQubit) {
+    }
+
+    bool SymbolicOperation::equals(const Operation& op, const Permutation& perm1, const Permutation& perm2) const {
+        if (!op.isSymbolicOperation() && !isStandardOperation()) {
+            return false;
+        }
+        if (isStandardOperation() && qc::StandardOperation::equals(op, perm1, perm2)) return true;
+
+        if (!op.isSymbolicOperation()) return false;
+        const auto& symOp = dynamic_cast<const SymbolicOperation&>(op);
+        for (std::size_t i = 0; i < symbolicParameter.size(); ++i) {
+            if (symbolicParameter[i].has_value() != symOp.symbolicParameter[i].has_value())
+                return false;
+
+            if (symbolicParameter[i].has_value())
+                return symbolicParameter[i].value() == symOp.symbolicParameter[i].value();
+        }
+        return true;
+    }
+
+    void SymbolicOperation::dumpOpenQASM([[maybe_unused]] std::ostream& of, [[maybe_unused]] const RegisterNames& qreg, [[maybe_unused]] const RegisterNames& creg) const {
+        throw QFRException("OpenQasm2.0 doesn't support parametrized gates!");
+    }
+    void SymbolicOperation::dumpQiskit([[maybe_unused]] std::ostream& of, [[maybe_unused]] const RegisterNames& qreg, [[maybe_unused]] const RegisterNames& creg, [[maybe_unused]] const char* anc_reg_name) const {
+        throw QFRException("Dumping Qiskit Circuit is not supported for parameterized gates.");
+    }
+
+    StandardOperation SymbolicOperation::getInstantiatedOperation(const VariableAssignment& assignment) const {
+        auto lambda = getInstantiation(getParameter(0), assignment);
+        auto phi    = getInstantiation(getParameter(1), assignment);
+        auto theta  = getInstantiation(getParameter(2), assignment);
+        return StandardOperation(nqubits, targets, type, lambda, phi, theta, startQubit);
+    }
+
+    // Instantiates this Operation
+    // Afterwards casting to StandardOperation can be done if assignment is total
+    void SymbolicOperation::instantiate(const VariableAssignment& assignment) {
+        parameter[0] = getInstantiation(getParameter(0), assignment);
+        symbolicParameter[0].reset();
+        parameter[1] = getInstantiation(getParameter(1), assignment);
+        symbolicParameter[1].reset();
+        parameter[2] = getInstantiation(getParameter(2), assignment);
+        symbolicParameter[2].reset();
```

Comment:
this could easily be written in a more generic fashion so that it does not depend on the number of parameters being 3.

Reply:
Right. Might have been during a late-night coding session.